File: crates/konedrived/src/quickxor.rs
//! QuickXorHash: the one content hash Microsoft Graph guarantees for files on
//! personal and business drives alike.
//!
//! The published algorithm XORs byte `k` of the content into a 160-bit
//! circular register at bit offset `(11 * k) mod 160`, then XORs the content's
//! length, little-endian, into the register's last 8 bytes. The offset depends
//! only on `k mod 160`, and XOR is linear, so every byte is first folded into
//! one of 160 lanes — one XOR per byte, whatever the file's size — and the
//! lanes are placed into the register once, at the end.

use base64::Engine;

const WIDTH_BITS: usize = 160;
const SHIFT: usize = 11;
/// Bytes in a hash.
pub const LEN: usize = 20;

#[derive(Clone)]
pub struct QuickXor {
    lanes: [u8; WIDTH_BITS],
    length: u64,
}

impl Default for QuickXor {
    fn default() -> Self {
        Self::new()
    }
}

impl QuickXor {
    pub fn new() -> Self {
        Self { lanes: [0; WIDTH_BITS], length: 0 }
    }

    pub fn update(&mut self, data: &[u8]) {
        let mut lane = (self.length % WIDTH_BITS as u64) as usize;
        for &byte in data {
            self.lanes[lane] ^= byte;
            lane += 1;
            if lane == WIDTH_BITS {
                lane = 0;
            }
        }
        self.length += data.len() as u64;
    }

    pub fn finish(&self) -> [u8; LEN] {
        let mut register = [0u8; LEN];
        for (lane, &value) in self.lanes.iter().enumerate() {
            if value == 0 {
                continue;
            }
            let offset = (lane * SHIFT) % WIDTH_BITS;
            for bit in 0..8 {
                if value >> bit & 1 == 1 {
                    let position = (offset + bit) % WIDTH_BITS;
                    register[position / 8] ^= 1 << (position % 8);
                }
            }
        }
        for (i, byte) in self.length.to_le_bytes().iter().enumerate() {
            register[LEN - 8 + i] ^= byte;
        }
        register
    }

    /// As Graph spells it in `file.hashes.quickXorHash`.
    pub fn finish_base64(&self) -> String {
        base64::engine::general_purpose::STANDARD.encode(self.finish())
    }
}

/// A `quickXorHash` as Graph spells it, or `None` if it is not one.
pub fn decode_base64(value: &str) -> Option<[u8; LEN]> {
    base64::engine::general_purpose::STANDARD.decode(value).ok()?.try_into().ok()
}

#[cfg(test)]
mod tests {
    use super::*;

    fn hash(data: &[u8]) -> String {
        let mut h = QuickXor::new();
        h.update(data);
        h.finish_base64()
    }

    /// A literal transcription of the published algorithm, one bit at a time:
    /// byte `k` goes into a 160-bit circular register at bit `(11 * k) mod 160`,
    /// and the length, little-endian, is XORed into the last 8 bytes. Nothing
    /// in it is shared with the implementation under test.
    fn reference(data: &[u8]) -> [u8; 20] {
        let mut register = [0u8; 20];
        for (k, &byte) in data.iter().enumerate() {
            let offset = (k * 11) % 160;
            for bit in 0..8 {
                if byte >> bit & 1 == 1 {
                    let position = (offset + bit) % 160;
                    register[position / 8] ^= 1 << (position % 8);
                }
            }
        }
        for (i, byte) in (data.len() as u64).to_le_bytes().iter().enumerate() {
            register[12 + i] ^= byte;
        }
        register
    }

    /// Deterministic noise, so a failure reproduces.
    fn noise(len: usize, seed: u64) -> Vec<u8> {
        let mut state = seed;
        (0..len)
            .map(|_| {
                state = state.wrapping_mul(6364136223846793005).wrapping_add(1442695040888963407);
                (state >> 56) as u8
            })
            .collect()
    }

    #[test]
    fn nothing_hashes_to_twenty_zero_bytes() {
        assert_eq!(hash(b""), "AAAAAAAAAAAAAAAAAAAAAAAAAAA=");
    }

    #[test]
    fn one_byte_lands_at_bit_zero_and_the_length_at_byte_twelve() {
        // 0x4a at offset 0 → byte 0 = 0x4a; length 1 → byte 12 = 0x01.
        assert_eq!(hash(&[0x4a]), "SgAAAAAAAAAAAAAAAQAAAAAAAAA=");
    }

    #[test]
    fn the_second_byte_lands_eleven_bits_further() {
        // 0xb5 → byte 0. 0xb4 = bits 2,4,5,7 → positions 13,15,16,18:
        // byte 1 gets bits 5,7 (0xa0), byte 2 bits 0,2 (0x05). Length 2 → byte 12.
        assert_eq!(hash(&[0xb5, 0xb4]), "taAFAAAAAAAAAAAAAgAAAAAAAAA=");
    }

    #[test]
    fn matches_the_bit_by_bit_transcription_on_every_length_up_to_a_thousand() {
        for len in 0..1000 {
            let data = noise(len, len as u64);
            let mut h = QuickXor::new();
            h.update(&data);
            assert_eq!(h.finish(), reference(&data), "length {len}");
        }
    }

    #[test]
    fn matches_the_transcription_on_a_megabyte() {
        let data = noise(1 << 20, 7);
        let mut h = QuickXor::new();
        h.update(&data);
        assert_eq!(h.finish(), reference(&data));
    }

    #[test]
    fn a_stream_in_pieces_hashes_like_the_whole() {
        let data = noise(100_003, 11);
        let mut whole = QuickXor::new();
        whole.update(&data);
        // Piece sizes that straddle the 160-byte period in every way.
        for sizes in [[1usize, 159, 160, 161], [7, 300, 13, 2], [160, 160, 160, 160]] {
            let mut pieces = QuickXor::new();
            let mut at = 0;
            let mut i = 0;
            while at < data.len() {
                let end = (at + sizes[i % sizes.len()]).min(data.len());
                pieces.update(&data[at..end]);
                at = end;
                i += 1;
            }
            assert_eq!(pieces.finish(), whole.finish(), "pieces {sizes:?}");
        }
    }

    #[test]
    fn base64_round_trips() {
        let h = {
            let mut h = QuickXor::new();
            h.update(b"konedrive");
            h
        };
        assert_eq!(decode_base64(&h.finish_base64()), Some(h.finish()));
        assert_eq!(decode_base64("not base64!"), None);
        assert_eq!(decode_base64("AAAA"), None, "three bytes are not a quickXorHash");
    }
}

File: tests/stress/quickxor.py
"""A pure-Python, stdlib-only port of Microsoft's QuickXorHash, the one content hash Graph
guarantees for files on personal and business drives alike. This is the same algorithm as
``crates/konedrived/src/quickxor.rs`` (read that file for the description); this module exists
only so ``stress_uploads.py`` can compute the hash of a local file without shelling out to Rust
or adding a dependency.

The update step is, byte for byte, ``lanes[k % 160] ^= data[k]`` for a running counter ``k`` — a
fold of the stream into 160 buckets. That is exactly a bytewise XOR of 160-byte blocks once the
stream is aligned to a 160-byte boundary, so ``update`` XORs 160-byte blocks together as big
integers (fast, in C) instead of looping one byte at a time in Python, which would be far too
slow for the 30-60 MiB files this tool uploads. ``finish`` (called once, on a short 160-byte
array) is cheap enough to stay a plain bit-by-bit loop, matching the Rust source directly.

Self-checked at import time against the three known-answer vectors from
``crates/konedrived/src/quickxor.rs``'s test suite, plus a differential check of the fast path
above against a bit-by-bit reference, across split points that straddle the 160-byte period. A
transcription mistake here should fail loudly the first time this module is imported, not
silently produce hashes that happen to agree with each other and disagree with OneDrive.
"""

from __future__ import annotations

import base64

WIDTH_BITS = 160
SHIFT = 11
LEN = 20


class QuickXor:
    """Incremental QuickXorHash. Call :meth:`update` as many times as convenient, then
    :meth:`finish` or :meth:`finish_base64` once."""

    __slots__ = ("_lanes_int", "length")

    def __init__(self) -> None:
        self._lanes_int = 0  # 160 lanes, packed as one big-endian integer of 160 bytes.
        self.length = 0

    def update(self, data: bytes) -> None:
        if not data:
            return
        start = self.length % WIDTH_BITS
        # Pad so the buffer starts at lane 0: the padding bytes are zero, so XOR-ing them in is
        # a no-op, and every real byte lands at the same lane it would one at a time.
        padded = bytes(start) + bytes(data)
        pad_tail = (-len(padded)) % WIDTH_BITS
        if pad_tail:
            padded += bytes(pad_tail)
        acc = 0
        for i in range(0, len(padded), WIDTH_BITS):
            acc ^= int.from_bytes(padded[i : i + WIDTH_BITS], "big")
        self._lanes_int ^= acc
        self.length += len(data)

    def finish(self) -> bytes:
        lanes = self._lanes_int.to_bytes(WIDTH_BITS, "big")
        register = bytearray(LEN)
        for lane, value in enumerate(lanes):
            if value == 0:
                continue
            offset = (lane * SHIFT) % WIDTH_BITS
            for bit in range(8):
                if (value >> bit) & 1:
                    position = (offset + bit) % WIDTH_BITS
                    register[position // 8] ^= 1 << (position % 8)
        length_bytes = (self.length & 0xFFFFFFFFFFFFFFFF).to_bytes(8, "little")
        for i, byte in enumerate(length_bytes):
            register[LEN - 8 + i] ^= byte
        return bytes(register)

    def finish_base64(self) -> str:
        return base64.b64encode(self.finish()).decode("ascii")


def hash_bytes(data: bytes) -> str:
    h = QuickXor()
    h.update(data)
    return h.finish_base64()


def hash_file(path, chunk_size: int = 4 * 1024 * 1024) -> str:
    h = QuickXor()
    with open(path, "rb") as f:
        while True:
            chunk = f.read(chunk_size)
            if not chunk:
                break
            h.update(chunk)
    return h.finish_base64()


def _reference(data: bytes) -> bytes:
    """A literal bit-by-bit transcription, independent of the fast path in ``update``. Used
    only by the self-test below."""
    register = bytearray(LEN)
    for k, byte in enumerate(data):
        offset = (k * SHIFT) % WIDTH_BITS
        for bit in range(8):
            if (byte >> bit) & 1:
                position = (offset + bit) % WIDTH_BITS
                register[position // 8] ^= 1 << (position % 8)
    length_bytes = (len(data) & 0xFFFFFFFFFFFFFFFF).to_bytes(8, "little")
    for i, byte in enumerate(length_bytes):
        register[LEN - 8 + i] ^= byte
    return bytes(register)


def _noise(length: int, seed: int) -> bytes:
    """The same small LCG the Rust tests use, for reproducible pseudo-random content."""
    state = seed
    out = bytearray(length)
    for i in range(length):
        state = (state * 6364136223846793005 + 1442695040888963407) & 0xFFFFFFFFFFFFFFFF
        out[i] = (state >> 56) & 0xFF
    return bytes(out)


def _self_test() -> None:
    # The three known-answer vectors from crates/konedrived/src/quickxor.rs.
    vectors = [
        (b"", "AAAAAAAAAAAAAAAAAAAAAAAAAAA="),
        (bytes([0x4A]), "SgAAAAAAAAAAAAAAAQAAAAAAAAA="),
        (bytes([0xB5, 0xB4]), "taAFAAAAAAAAAAAAAgAAAAAAAAA="),
    ]
    for data, expected in vectors:
        got = hash_bytes(data)
        if got != expected:
            raise AssertionError(
                f"quickxor.py self-test failed for {data!r}: got {got}, want {expected} "
                "(check this port against crates/konedrived/src/quickxor.rs)"
            )

    # Differential check of the fast (block-XOR) update() against the bit-by-bit reference,
    # across lengths and split points that straddle the 160-byte period in different ways —
    # the thing most likely to be wrong in a "fold into 160-byte blocks" rewrite.
    for length in (0, 1, 2, 3, 159, 160, 161, 319, 320, 321, 733):
        data = _noise(length, seed=length + 1)
        want = _reference(data)
        for splits in ([length], [1, length - 1], [159, 160, length - 319], [7, 300, 13]):
            splits = [s for s in splits if s > 0]
            if sum(splits) != length:
                continue
            h = QuickXor()
            at = 0
            for s in splits:
                h.update(data[at : at + s])
                at += s
            got = h.finish()
            if got != want:
                raise AssertionError(
                    f"quickxor.py self-test failed: length {length}, splits {splits}: "
                    f"got {got.hex()}, want {want.hex()}"
                )


_self_test()


if __name__ == "__main__":
    print("quickxor.py: self-test passed.")
